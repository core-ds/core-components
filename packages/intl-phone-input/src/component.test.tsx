import React from 'react';
import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { IntlPhoneInput } from './index';

describe('IntlPhoneInput', () => {
    const testId = 'intl-phone-input';

    it('should match snapshot', () => {
        const { container } = render(
            <IntlPhoneInput value='+7 111 111 11 11' onChange={jest.fn()} />,
        );

        expect(container).toMatchSnapshot();
    });

    it('should unmount without errors', () => {
        const { unmount } = render(
            <IntlPhoneInput value='+7 111 111 11 11' onChange={jest.fn()} />,
        );

        expect(unmount).not.toThrowError();
    });

    it('should call `onChange` callback after input was changed with dial code of country without priority', async () => {
        const onChange = jest.fn();
        render(<IntlPhoneInput value='' onChange={onChange} dataTestId={testId} />);

        const input = await screen.getByDisplayValue('');

        fireEvent.change(input, { target: { value: '+54' } });

        expect(onChange).toHaveBeenCalled();
        expect(onChange).toHaveBeenCalledWith('+54');
    });

    it('should call `onChange` callback after input was changed with dial code of country from NANP', async () => {
        const onChange = jest.fn();
        render(<IntlPhoneInput value='' onChange={onChange} dataTestId={testId} />);

        const input = await screen.getByDisplayValue('');

        fireEvent.change(input, { target: { value: '+1868' } });

        expect(onChange).toHaveBeenCalled();
        expect(onChange).toHaveBeenCalledWith('+1868');
    });

    it('should call `onChange` callback after input was changed with whole russian number', async () => {
        const onChange = jest.fn();
        render(<IntlPhoneInput value='' onChange={onChange} dataTestId={testId} />);

        const input = await screen.getByDisplayValue('');
        fireEvent.change(input, { target: { value: '+74957888878' } });

        expect(onChange).toHaveBeenCalled();
        expect(onChange).toHaveBeenCalledWith('+74957888878');
    });

    it('should have default country flag icon', () => {
        const { container } = render(<IntlPhoneInput value='' onChange={jest.fn()} />);

        const flagComponent = container.querySelector('.flagIcon');

        expect(flagComponent).toHaveClass('ru');
    });

    it('should have passed country flag icon', () => {
        const { container } = render(
            <IntlPhoneInput value='' onChange={jest.fn()} defaultCountryIso2='az' />,
        );

        const flagComponent = container.querySelector('.flagIcon');

        expect(flagComponent).toHaveClass('az');
    });

    it('should set new country flag icon from props', async () => {
        const { container } = render(<IntlPhoneInput value='+61' onChange={jest.fn()} />);

        await waitFor(() => expect(container.querySelector('.flagIcon')).toHaveClass('au'));
    });

    it('should call `onChange` callback after select was changed', async () => {
        const onChange = jest.fn();

        const { container, getAllByRole } = render(<IntlPhoneInput value='' onChange={onChange} />);

        const flagComponent = container.querySelector('.flagIcon');

        fireEvent.click(flagComponent as HTMLSpanElement);
        const option = await waitFor(() => getAllByRole('option')[0]);
        fireEvent.click(option);

        waitFor(() => {
            expect(onChange).toHaveBeenCalled();
        });
    });

    it('should focus on input after select was changed', async () => {
        const { container, getAllByRole } = render(
            <IntlPhoneInput value='+7' onChange={() => null} dataTestId={testId} />,
        );
        const input = await screen.findByDisplayValue('+7');
        const flagComponent = container.querySelector('.flagIcon');

        fireEvent.click(flagComponent as HTMLSpanElement);
        const option = await waitFor(() => getAllByRole('option')[0]);
        fireEvent.click(option);

        waitFor(() => {
            expect(document.activeElement).toBe(input);
        });
    });

    it('should call `onCountryChange` callback after country was changed', async () => {
        const onCountryChange = jest.fn();
        const { container, getAllByRole } = render(
            <IntlPhoneInput
                value=''
                onChange={() => null}
                onCountryChange={onCountryChange}
                dataTestId={testId}
            />,
        );

        const flagComponent = container.querySelector('span[class^=flagIcon]');
        fireEvent.click(flagComponent as HTMLSpanElement);
        const option = await waitFor(() => getAllByRole('option')[0]);
        fireEvent.click(option);

        await waitFor(() => expect(onCountryChange).toHaveBeenCalledTimes(1));
    });

    it('should call `onCountryChange` callback after input was changed', async () => {
        const onCountryChange = jest.fn();
        render(
            <IntlPhoneInput
                value=''
                onChange={() => null}
                dataTestId={testId}
                onCountryChange={onCountryChange}
                defaultCountryIso2='ru'
            />,
        );
        const input = await screen.getByDisplayValue('');

        fireEvent.change(input, { target: { value: '+998 12 345 67 89', selectionStart: 0 } });

        expect(onCountryChange).toBeCalledWith('UZ');
        expect(onCountryChange).toHaveBeenCalledTimes(1);
    });

    it('should remove country code', async () => {
        const onChange = jest.fn();
        render(<IntlPhoneInput value='+7' onChange={onChange} />);

        const input = await screen.findByDisplayValue('+7');
        fireEvent.change(input, { target: { value: '+' } });

        waitFor(() => {
            expect(onChange).toHaveBeenCalledWith('+');
            expect(input).toHaveValue('+');
        });
    });

    it('should not remove country code', async () => {
        const onChange = jest.fn();
        render(<IntlPhoneInput value='+7' onChange={onChange} clearableCountryCode={false} />);

        const input = await screen.findByDisplayValue('+7');
        fireEvent.change(input, { target: { value: '+' } });

        waitFor(() => {
            expect(onChange).not.toHaveBeenCalled();
            expect(input).toHaveValue('+7');
        });
    });

    it('should not show country select when one country is available', async () => {
        const countries = [
            {
                areaCodes: null,
                dialCode: '7',
                iso2: 'ru',
                name: 'Россия',
                priority: 0,
            },
        ];
        const { queryByTestId } = render(
            <IntlPhoneInput value='+7' countries={countries} onChange={() => null} />,
        );

        const countrySelect = queryByTestId('countries-select');

        expect(countrySelect).toBeNull();
    });

    it('should show empty country', async () => {
        const onCountryChange = jest.fn();
        render(
            <IntlPhoneInput
                value=''
                onChange={() => null}
                dataTestId={testId}
                onCountryChange={onCountryChange}
                canBeEmptyCountry={true}
                defaultCountryIso2='ru'
            />,
        );

        const input = screen.getByDisplayValue('');

        fireEvent.change(input, { target: { value: '+', selectionStart: 0 } });

        const icons = screen.getAllByRole('img');

        expect(icons[0]).toHaveClass('emptyCountryIcon');
        expect(onCountryChange).toBeCalledWith(undefined);
    });

    it('should call `onChange` with value "+7" when type "8" in empty field with `ruNumberPriority`', async () => {
        const onChange = jest.fn();
        render(
            <IntlPhoneInput
                value=''
                onChange={onChange}
                dataTestId={testId}
                ruNumberPriority={true}
                defaultCountryIso2='ru'
            />,
        );

        const input = screen.getByDisplayValue('');

        userEvent.type(input, '8');

        await waitFor(() => {
            expect(onChange).toBeCalledWith('+7');
        });
    });

    it('should call `onChange` with value "+7 9" when type "9" in empty field with `ruNumberPriority`', async () => {
        const onChange = jest.fn();
        render(
            <IntlPhoneInput
                value=''
                onChange={onChange}
                dataTestId={testId}
                ruNumberPriority={true}
                defaultCountryIso2='ru'
            />,
        );

        const input = screen.getByDisplayValue('');

        userEvent.type(input, '9');

        await waitFor(() => {
            expect(onChange).toBeCalledWith('+7 9');
        });
    });

    it('should call `onChange` with value "+7" when type "7" in empty field with `ruNumberPriority`', async () => {
        const onChange = jest.fn();
        render(
            <IntlPhoneInput
                value=''
                onChange={onChange}
                dataTestId={testId}
                ruNumberPriority={true}
                defaultCountryIso2='ru'
            />,
        );

        const input = screen.getByDisplayValue('');

        userEvent.type(input, '7');

        await waitFor(() => {
            expect(onChange).toBeCalledWith('+7');
        });
    });

    it('should not render country select when use `hideCountrySelect`', () => {
        const onChange = jest.fn();
        render(
            <IntlPhoneInput
                value=''
                onChange={onChange}
                dataTestId={testId}
                hideCountrySelect={true}
                defaultCountryIso2='ru'
            />,
        );

        const countrySelect = screen.queryByTestId('countries-select');

        expect(countrySelect).toBeNull();
    });

    it('should show "+" when clear value', async () => {
        const onChange = jest.fn();
        render(<IntlPhoneInput value='+7' onChange={onChange} clear={true} />);

        const clearButton = await screen.findByLabelText('Очистить');

        expect(clearButton).toBeInTheDocument();

        fireEvent.click(clearButton);

        await waitFor(() => {
            expect(onChange).toBeCalledWith('+');
        });
    });
});
