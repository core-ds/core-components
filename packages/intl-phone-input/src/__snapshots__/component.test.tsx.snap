// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IntlPhoneInput should match snapshot 1`] = `
<div>
  <div
    aria-haspopup="listbox"
    aria-owns="downshift-1-menu"
    class="component"
    role="combobox"
    tabindex="-1"
  >
    <div
      class="component component m m block hasLeftAddons"
    >
      <div
        class="inner inner filled"
      >
        <div
          class="addons leftAddons addons"
        >
          <div
            class="component"
          >
            <div
              aria-haspopup="listbox"
              aria-owns="downshift-3-menu"
              class="component"
              data-test-id="countries-select"
              role="combobox"
              tabindex="-1"
            >
              <div
                class="component m"
              >
                <div
                  aria-controls="downshift-3-menu"
                  aria-labelledby="downshift-3-label"
                  class="inner"
                  id="downshift-2-input"
                  tabindex="0"
                >
                  <span
                    class="flagIconContainer"
                  >
                    <span
                      class="flagIcon"
                      data-test-id="flag-icon-ru"
                    >
                      <svg
                        fill="none"
                        height="24"
                        width="24"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M0 0h24v24H0z"
                        />
                        <path
                          d="M22.2 4.1H1.8C.8 4.1 0 5 0 6v3.5h24V5.9c0-1-.8-1.8-1.8-1.8Z"
                          fill="#F5F5F5"
                        />
                        <path
                          d="M0 18C0 19 .8 20 1.8 20h20.4c1 0 1.8-.8 1.8-1.8v-3.5H0v3.5Z"
                          fill="#FF4B55"
                        />
                        <path
                          d="M24 9.4H0v5.2h24V9.4Z"
                          fill="#41479B"
                        />
                      </svg>
                    </span>
                  </span>
                  <svg
                    class="arrow"
                    fill="currentColor"
                    focusable="false"
                    height="24"
                    role="img"
                    viewBox="0 0 24 24"
                    width="24"
                  >
                    <path
                      d="M12 14L6.5 8.5 5 10l7 7 7-7-1.5-1.5L12 14z"
                    />
                  </svg>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="inputWrapper"
        >
          <div
            class="input"
          >
            <input
              aria-autocomplete="list"
              aria-controls="downshift-1-menu"
              aria-labelledby="downshift-1-label"
              autocomplete="off"
              class="input input"
              id="downshift-0-input"
              tabindex="0"
              type="tel"
              value="+7 111 111 11 11"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
